post: Ensuring secure values by private keys in AWS (KMS, SSM, Secrets Manager) (#78)
cloudx-labs/posts: posts/frangeris/03.ensure-secure-private-keys-using-aws-kms.md

---
title: Ensuring secure values by private keys in AWS (KMS, SSM, Secrets Manager)
published: false
description: Secure way of store and access sensible values on AWS
tags: "aws, private keys security, serverless security, cloud security"
cover_image: ./assets/03.cover.png
---

## Ensuring secure values by private keys in AWS (KMS, SSM, Secrets Manager)

Private keys, credentials access and sensitive information such as passwords or any kind of confidential data are essentials in applications logic when interacting with other services or even databases, storing those values can be risky if proper security measures are not in place.

In this post we will dig into one way to accomplish secure access to sensible information in AWS environments. For simplicity we will be using:

> - A ready to use [AWS account](https://aws.amazon.com/)

Reviewer: ```suggestion
- A ready to use [AWS account](https://aws.amazon.com).
```

> - [Serverless Framework](https://www.serverless.com/framework/docs/getting-started) - to delegate the CloudFormation creation and deployment

Reviewer: ```suggestion
- [Serverless Framework](https://www.serverless.com/framework/docs/getting-started) - to delegate the CloudFormation creation and deployment.
```

By using `npx` command we get access to serverless executable, select the basic to keep it simple, there are options for different types of uses cases, even different languages options in case `Node.js` is not the language

![Run npx serverless command to start](./assets/03.image.1.png)

Some operation used on the lambda will require some permissions later, due to this, its recommended to include and enable those permissions before the IAM role of the lambda gets created, here is what permissions we need that the lambda has:

```yaml
provider:
  # ...
  iam:
    role:
      statements:
        - Effect: Allow
          Action:
            - "s3:*"
          Resource: "*"
        - Effect: Allow
          Action:
            - "cloudwatch:*"
            - "logs:*"
          Resource: "*"
        - Effect: Allow
          Action:
            - "ssm:GetParameter"
            - "ssm:GetParametersByPath"
          Resource: "*"
        - Effect: Allow
          Action:
            - "secretsmanager:GetSecretValue"
          Resource: "*"
        - Effect: Allow
          Action:
            - "kms:Decrypt"
          Resource: "*"
```

> Then, with you [aws account](https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-files.html) already setted up, let's deploy it

Reviewer: ```suggestion
Then, with you [aws account](https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-files.html) already setted up, let's deploy it.
```

```bash
npx sls deploy
```

> 🎉 After a correct deployment, the code should be created successfully and ready to use, now we need to configure the most important part, security,

Reviewer: ```suggestion
🎉 After a correct deployment, the code should be created successfully and ready to use, now we need to configure the most important part, security.
```

## Security

Permission will be managed via [IAM policies](https://docs.aws.amazon.com/IAM/latest/UserGuide/access_policies.html) with the objective of _only allows the lambda to get and decrypt the secured value_, the approach used will be based on: **prevent IAM entities from accessing the KMS key and allow the root user account to manage it** (also preventing the root user account from losing access to the KMS key).

This "protection" layer of our sensitive value is based on the composition of two AWS services, [Key Management Service](https://aws.amazon.com/kms/) is the one responsible for creating the private key which will be used to encrypt our value, then [Secret Manager](https://aws.amazon.com/es/secrets-manager/) or [System Manager: Parameters Store](https://aws.amazon.com/systems-manager/), both services allows us to accomplish the same functionality but differ on some caveats, _the chosen one_ will be used for saving the secured encrypted value, here is a quick graphic representation of the implementation:

> ![Diagram that explain how services are used](./assets/03.image.2.png)

Reviewer: ```suggestion
![A diagram that explains how services are used](./assets/03.image.2.png)
```

```bash
npx sls info --verbose
```

> --plaintext (data to be encrypted) is required to be base64 text or a blob from a plain text file

```bash
aws kms encrypt \
    --region <same region used by serverless> \
    --key-id <key id from output (MyPrivateKeyId)> \
    --plaintext "<base64 of value> or <fileb://path>" \
    --output text \
    --query CiphertextBlob

aws kms encrypt --profile skuadme --region us-east-1 --key-id  alias/myprivatekey --plaintext "bXkgcHJlY2lvdXMgdmFsdWU=" --output text --query CiphertextBlob
```

🎉 You should be able to get a base64 output from the command above with the value encrypted, now what next?

### Store the values

We need to save those values inside a secure service, either AWS System Manager (using Parameter Store) or AWS Secret Manager creating directly a key/value, why not try both?

#### For Parameter Store

For reusability use the previous encrypted output directly with the same command, getting as result:

```bash
aws ssm put-parameter \
  --name "my-parameter-name" \
  --type "SecureString" \

>   --value "$(aws kms encrypt --region us-east-1 --key-id <key id> --plaintext "<value>" --output text --query CiphertextBlob | base64 --decode)" \

Reviewer: ```suggestion
  --value "$(aws kms encrypt --region us-east-1 --key-id <key_id> --plaintext "<value>" --output text --query CiphertextBlob | base64 --decode)" \
```

  --region us-east-1

aws ssm put-parameter \
  --profile skuadme \
  --name "my-parameter-name" \
  --type "SecureString" \
  --value "$(aws kms encrypt --profile skuadme --region us-east-1 --key-id alias/myprivatekey --plaintext "bXkgcHJlY2lvdXMgdmFsdWU=" --output text --query CiphertextBlob)" \
  --region us-east-1

> 

Reviewer: ```suggestion
```

```

If everything works, you should receive json payload like this, which means it works, and the value was created successfully 👏🏽

> > Access via AWS Systems Manager > Parameter Store to see it.

Reviewer: ```suggestion
> Access via AWS Systems Manager
> Parameter Store to see it.
```
